accelerate-llvm/src/Data/Array/Accelerate/LLVM/CodeGen/Profile.hs

```diff
@@ -0,0 +1,192 @@
+{-# LANGUAGE ForeignFunctionInterface #-}
+{-# LANGUAGE OverloadedStrings        #-}
+{-# LANGUAGE TypeApplications         #-}
+-- |
+-- Module      : Data.Array.Accelerate.LLVM.CodeGen.Profile
+-- Copyright   : [2015..2020] The Accelerate Team
+-- License     : BSD3
+--
+-- Maintainer  : Trevor L. McDonell <trevor.mcdonell@gmail.com>
+-- Stability   : experimental
+-- Portability : non-portable (GHC extensions)
+--
+
+module Data.Array.Accelerate.LLVM.CodeGen.Profile (
+
+  zone_begin,
+  zone_end,
+
+  alloc_srcloc,
+  alloc_srcloc_name,
+  zone_begin_alloc,
+
+) where
+
+import LLVM.AST.Type.AddrSpace
+import LLVM.AST.Type.Constant
+import LLVM.AST.Type.Downcast
+import LLVM.AST.Type.Function
+import LLVM.AST.Type.Global
+import LLVM.AST.Type.Instruction
+import LLVM.AST.Type.Name
+import LLVM.AST.Type.Operand
+import LLVM.AST.Type.Representation
+import qualified LLVM.AST.Constant                                  as Constant
+import qualified LLVM.AST.Global                                    as LLVM
+import qualified LLVM.AST.Linkage                                   as LLVM
+import qualified LLVM.AST.Type                                      as LLVM
+
+import Data.Array.Accelerate.LLVM.CodeGen.Base
+import Data.Array.Accelerate.LLVM.CodeGen.Constant
+import Data.Array.Accelerate.LLVM.CodeGen.IR
+import Data.Array.Accelerate.LLVM.CodeGen.Monad
+
+import Data.Array.Accelerate.Sugar.Elt
+import Data.Array.Accelerate.Debug.Internal                         ( profilingIsEnabled, SrcLoc, Zone )
+
+import Control.Monad
+import Data.Char
+
+
+call' :: GlobalFunction args t -> CodeGen arch (Operands t)
+call' f = call f [NoUnwind, NoDuplicate]
+
+global_string :: String -> CodeGen arch (Name (Ptr Word8), Word64)
+global_string cs = do
+  let l  = fromIntegral (length cs)
+  --
+  nm <- freshGlobalName
+  _  <- declare $ LLVM.globalVariableDefaults
+    { LLVM.name        = downcast nm
+    , LLVM.isConstant  = True
+    , LLVM.linkage     = LLVM.Private
+    , LLVM.type'       = LLVM.ArrayType l LLVM.i8
+    , LLVM.unnamedAddr = Just LLVM.GlobalAddr
+    , LLVM.initializer = Just $ Constant.Array LLVM.i8 [ Constant.Int 8 (toInteger (ord c)) | c <- cs ]
```

Comment:
This would require the caller (who may not be aware of this since this function will likely be called indirectly as part of another function) to always manually terminate every `String` with a null byte, right? Would it make sense to move that responsibility to this function instead?

Reply:
Actually the strings don't have to be null terminated (at least for any function that also takes the length). I certainly am not adding the null terminator myself and it doesn't appear in the generated code. We might as well add it though.

Reply:
Yeah in that `___tracy_source_location_data` struct there's no string length defined anywhere, so they probably expect the strings to be null terminated. With these globals there's probably a decent chance that there will be a null byte following them because of alignment, but there wouldn't be a guarantee that there is (or that the compiler is using null bytes for padding).

Reply:
I expect they would `sizeof(struct)` and that will give you the right size, but either way you're right it's better we include it 👍